server/src/main/java/org/elasticsearch/search/query/QuerySearchResult.java

```diff
@@ -330,21 +329,31 @@ public void readFromWithId(ShardSearchContextId id, StreamInput in) throws IOExc
             }
         }
         setTopDocs(readTopDocs(in));
-        if (hasAggs = in.readBoolean()) {
-            aggregations = DelayableWriteable.delayed(InternalAggregations::readFrom, in);
-        }
-        if (in.readBoolean()) {
-            suggest = new Suggest(in);
-        }
-        searchTimedOut = in.readBoolean();
-        terminatedEarly = in.readOptionalBoolean();
-        profileShardResults = in.readOptionalWriteable(ProfileShardResult::new);
-        hasProfileResults = profileShardResults != null;
-        serviceTimeEWMA = in.readZLong();
-        nodeQueueSize = in.readInt();
-        if (in.getVersion().onOrAfter(Version.V_7_10_0)) {
-            setShardSearchRequest(in.readOptionalWriteable(ShardSearchRequest::new));
-            setRescoreDocIds(new RescoreDocIds(in));
+        hasAggs = in.readBoolean();
+        boolean success = false;
+        try {
+            if (hasAggs) {
+                aggregations = DelayableWriteable.delayed(InternalAggregations::readFrom, in);
+            }
+            if (in.readBoolean()) {
+                suggest = new Suggest(in);
+            }
+            searchTimedOut = in.readBoolean();
+            terminatedEarly = in.readOptionalBoolean();
+            profileShardResults = in.readOptionalWriteable(ProfileShardResult::new);
+            hasProfileResults = profileShardResults != null;
+            serviceTimeEWMA = in.readZLong();
+            nodeQueueSize = in.readInt();
+            if (in.getVersion().onOrAfter(Version.V_7_10_0)) {
+                setShardSearchRequest(in.readOptionalWriteable(ShardSearchRequest::new));
+                setRescoreDocIds(new RescoreDocIds(in));
+            }
+            success = true;
+        } finally {
+            if (success == false) {
```

Comment:
I we want to clean the aggregations only in case of an error, would be better to do it in a `catch` block, instead of creating an additional variable and tracking its status in `finally`?

Reply:
I figured this was the approach we used in most other places and it helps in so far (relative to catching) that we wouldn't leak here if we hit an error (most likely by tripping an assertion) and thus make test debugging easier.

Reply:
Just to add that not only does this pattern deal with the `Error` case without having to `catch Throwable` but it's also robust against future changes that introduce an early exit somewhere (e.g. a try-catch-log-and-return with smaller scope). Seems unlikely here but it's a good pattern anyway, I much prefer to signal the change of ownership explicitly like this than to do it implicitly in an exception handler.